core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -0,0 +1,180 @@
+   * Calling thread might be interrupted by this manager,
+   * if it found to be waiting for the shared lock under when:
+   *  - backup is exiting grace-cycle and entering the lock permanently
+   *  - backup is in progress
+   *
+   * @return the lock resource
+   * @throws InterruptedException
+   */
+  public LockResource lockShared() throws InterruptedException {
+    // Do not allow taking shared lock during safe-mode.
+    if (mSafeMode) {
+      String safeModeMsg = String.format(
+          "Cannot acquire lock shared during backup safe-mode. "
+              + "Please configure %s as 'false' in order to disable the backup safe-mode.",
+          PropertyKey.Name.MASTER_BACKUP_STATE_LOCK_SAFE_MODE);
+      throw new IllegalStateException(safeModeMsg);
+    }
+    // Register thread for interrupt cycle.
+    mSharedWaitersAndHolders.add(Thread.currentThread());
+    // Grab the lock interruptibly.
+    mStateLock.readLock().lockInterruptibly();
+    // Return the resource.
+    // Register an action to remove the thread from holders registry before releasing the lock.
+    return new LockResource(mStateLock.readLock(), false, false, () -> {
+      mSharedWaitersAndHolders.remove(Thread.currentThread());
+    });
+  }
+
+  /**
+   * Locks the state exclusively.
+   *
+   * @param lockOptions exclusive lock options
+   * @return the lock resource
+   * @throws TimeoutException if locking times out
+   * @throws InterruptedException if interrupting during locking
+   */
+  public LockResource lockExclusive(StateLockOptions lockOptions)
+      throws TimeoutException, InterruptedException {
+    // Run the grace cycle.
+    boolean lockAcquired = false;
+    if (lockOptions.waitGraceCycle()) {
+      long deadlineMs = System.currentTimeMillis() + lockOptions.getGraceCycleTimeoutMs();
+      while (System.currentTimeMillis() < deadlineMs) {
+        if (mStateLock.writeLock().tryLock(lockOptions.getGraceCycleTryMs(), TimeUnit.MILLISECONDS)) {
+          lockAcquired = true;
+          break;
+        } else {
+          long remainingWaitMs = deadlineMs - System.currentTimeMillis();
+          if (remainingWaitMs > 0) {
+            Thread.sleep(Math.min(lockOptions.getGraceCycleSleepMs(), remainingWaitMs));
+          }
+        }
+      }
+    }
+    if (lockAcquired) { // Lock was acquired within grace-cycle.
+      scheduleInterruptCycle(lockOptions.getInterruptWaitersInternalMs());
+    } else { // Lock couldn't be acquired by grace-cycle.
+      if (!lockOptions.lockAfterGraceCycle()) {
+        throw new TimeoutException(
+                ExceptionMessage.STATE_LOCK_TIMED_OUT.getMessage(lockOptions.getGraceCycleTimeoutMs()));
+      }
+      if (lockOptions.interruptCycleAfterGraceCycle()) {
+        // Schedule interrupt cycle before entering the lock because it might wait in the queue.
+        scheduleInterruptCycle(lockOptions.getInterruptWaitersInternalMs());
+      }
+      // Own the lock.
+      mStateLock.writeLock().lock();
```

Comment:
This could block for a long time, and thus block other RPCs (including heartbeats) for a long time?

Reply:
Yes but interrupt cycle is activated before this point.